src/paginationCalls/pageDescriptor.ts

```diff
@@ -103,6 +116,107 @@ export class PageDescriptor implements Descriptor {
     return stream;
   }
 
+  /**
+   * Create an async iterable which can be recursively called for data on-demand.
+   */
+  asyncIterate(
+    apiCall: GaxCall,
+    request: RequestType,
+    options: CallSettings
+  ): AsyncIterable<{} | undefined> {
+    const iterable = this.createIterator(options);
+    const funcPromise =
+      typeof apiCall === 'function' ? Promise.resolve(apiCall) : apiCall;
+    funcPromise
+      .then((func: GaxCall) => {
+        this.resolveParams(request, func, options);
+      })
+      .catch(error => {
+        throw new Error(error);
+      });
+    return iterable;
+  }
+
+  createIterator(options: CallSettings): AsyncIterable<{} | undefined> {
+    const responsePageTokenFieldName = this.responsePageTokenField;
+    const requestPageTokenFieldName = this.requestPageTokenField;
+    let resolveRequest = this.resolveRequest;
+    let resolveFunction = this.resolveFunction;
+    const asyncIterable = {
+      [Symbol.asyncIterator]() {
+        const funcPromise = new Promise((resolve, reject) => {
+          resolveFunction = resolve;
+        });
+        const requestPromise = new Promise((resolve, reject) => {
+          resolveRequest = resolve;
+        });
+        const cache: Array<{}> = [];
+        let nextPageRequest: RequestType | null = {};
+        let firstCall = true;
+        return {
+          async next() {
+            const ongoingCall = new call.OngoingCallPromise(options.promise);
+
+            const func = (await funcPromise) as SimpleCallbackFunction;
+            const request = (await requestPromise) as RequestType;
+            if (firstCall) {
+              ongoingCall.call(func, request);
+              const [
+                response,
+                nextRequest,
+                rawresponse,
+              ] = await ongoingCall.promise;
+              const pageToken = (response as ResponseType)
+                .responsePageTokenFieldName;
+              if (pageToken) {
+                nextPageRequest = Object.assign({}, request);
+                nextPageRequest[requestPageTokenFieldName] = pageToken;
+              }
+              const responses = (response as ResponseType).resourceField;
+              cache.push(...((responses as unknown) as Iterable<{}>));
```

Comment:
Are you sure that the cache will store `Iterable`s?

Reply:
so I tried `cache.push(...responses)` which should work in our case, but it complains about responses do not have `[Symbol.Iterator]()` and block the compile. Not sure why. Technically I think cache should just store object {value, done}

Reply:
after adding helper function, the problem goes away, so fixed it :)